File: Python/dinic.py
from collections import deque

class Dinic:
    def __init__(self, n):
        self.n = n
        self.graph = [[] for _ in range(n)]
        self.level = []

    def add_edge(self, u, v, capacity):
        # Forward edge: [to, capacity, reverse_edge_index]
        self.graph[u].append([v, capacity, len(self.graph[v])])
        # Backward edge with 0 initial capacity
        self.graph[v].append([u, 0, len(self.graph[u]) - 1])

    def bfs(self, s, t):
        """Builds the level graph using BFS."""
        self.level = [-1] * self.n
        self.level[s] = 0
        queue = deque([s])
        while queue:
            u = queue.popleft()
            for v, cap, rev_idx in self.graph[u]:
                if cap > 0 and self.level[v] == -1:
                    self.level[v] = self.level[u] + 1
                    queue.append(v)
        return self.level[t] != -1

    def dfs(self, u, t, flow, ptr):
        """Finds blocking flow using DFS."""
        if u == t or flow == 0:
            return flow
        
        for i in range(ptr[u], len(self.graph[u])):
            v, cap, rev_idx = self.graph[u][i]
            if self.level[v] == self.level[u] + 1 and cap > 0:
                pushed = self.dfs(v, t, min(flow, cap), ptr)
                if pushed > 0:
                    self.graph[u][i][1] -= pushed
                    self.graph[v][rev_idx][1] += pushed
                    return pushed
            ptr[u] += 1  # Optimization: mark edge as unusable for current level
        return 0

    def max_flow(self, s, t):
        max_f = 0
        while self.bfs(s, t):
            ptr = [0] * self.n
            while True:
                pushed = self.dfs(s, t, float('inf'), ptr)
                if pushed == 0:
                    break
                max_f += pushed
        return max_f

# Example usage
if __name__ == "__main__":
    dinic = Dinic(6)
    dinic.add_edge(0, 1, 10)
    dinic.add_edge(0, 2, 10)
    dinic.add_edge(1, 3, 4)
    dinic.add_edge(1, 4, 8)
    dinic.add_edge(1, 2, 2)
    dinic.add_edge(2, 4, 9)
    dinic.add_edge(3, 5, 10)
    dinic.add_edge(4, 3, 6)
    dinic.add_edge(4, 5, 10)

    source = 0
    sink = 5
    print("Maximum flow:", dinic.max_flow(source, sink))

File: TypeScript/dinic.ts
type Edge2 = {
    to: number;
    capacity: number;
    revIdx: number;
};

class Dinic {
    private n: number;
    private graph: Edge2[][];
    private level: number[] = [];

    constructor(n: number) {
        this.n = n;
        this.graph = Array.from({ length: n }, () => []);
    }

    addEdge2(u: number, v: number, capacity: number): void {
        // Forward edge2
        this.graph[u].push({ to: v, capacity, revIdx: this.graph[v].length });
        // Backward edge2
        this.graph[v].push({ to: u, capacity: 0, revIdx: this.graph[u].length - 1 });
    }

    private bfs(s: number, t: number): boolean {
        this.level = new Array(this.n).fill(-1);
        this.level[s] = 0;
        const queue: number[] = [s];

        while (queue.length > 0) {
            const u = queue.shift()!;
            for (const edge2 of this.graph[u]) {
                if (edge2.capacity > 0 && this.level[edge2.to] === -1) {
                    this.level[edge2.to] = this.level[u] + 1;
                    queue.push(edge2.to);
                }
            }
        }
        return this.level[t] !== -1;
    }

    private dfs(u: number, t: number, flow: number, ptr: number[]): number {
        if (u === t || flow === 0) return flow;

        for (; ptr[u] < this.graph[u].length; ptr[u]++) {
            const edge2 = this.graph[u][ptr[u]];
            if (this.level[edge2.to] === this.level[u] + 1 && edge2.capacity > 0) {
                const pushed = this.dfs(edge2.to, t, Math.min(flow, edge2.capacity), ptr);
                if (pushed > 0) {
                    edge2.capacity -= pushed;
                    this.graph[edge2.to][edge2.revIdx].capacity += pushed;
                    return pushed;
                }
            }
        }
        return 0;
    }

    maxFlow(s: number, t: number): number {
        let maxF = 0;
        while (this.bfs(s, t)) {
            const ptr = new Array(this.n).fill(0);
            while (true) {
                const pushed = this.dfs(s, t, Infinity, ptr);
                if (pushed === 0) break;
                maxF += pushed;
            }
        }
        return maxF;
    }
}

// Example usage
const dinic = new Dinic(6);
dinic.addEdge2(0, 1, 10);
dinic.addEdge2(0, 2, 10);
dinic.addEdge2(1, 3, 4);
dinic.addEdge2(1, 4, 8);
dinic.addEdge2(1, 2, 2);
dinic.addEdge2(2, 4, 9);
dinic.addEdge2(3, 5, 10);
dinic.addEdge2(4, 3, 6);
dinic.addEdge2(4, 5, 10);

const source = 0;
const sink = 5;
console.log("Maximum flow:", dinic.maxFlow(source, sink));
